tests/data/test_data_pipeline.py

```diff
@@ -0,0 +1,750 @@
+from flash.data.data_pipeline import _StageOrchestrator, DataPipeline
+from flash.data.process import Postprocess, Preprocess
+
+
+class DummyDataset(torch.utils.data.Dataset):
+
+    def __getitem__(self, index: int) -> Tuple[torch.Tensor, torch.Tensor]:
+        return torch.rand(1), torch.rand(1)
+
+    def __len__(self) -> int:
+        return 5
+
+
+class CustomModel(Task):
+
+    def __init__(self, postprocess: Optional[Postprocess] = None):
+        super().__init__(model=torch.nn.Linear(1, 1), loss_fn=torch.nn.MSELoss())
+        self._postprocess = postprocess
+
+    def train_dataloader(self) -> Any:
+        return DataLoader(DummyDataset())
+
+
+class CustomDataModule(DataModule):
+
+    def __init__(self):
+        super().__init__(
+            train_ds=DummyDataset(),
+            valid_ds=DummyDataset(),
+            test_ds=DummyDataset(),
+            predict_ds=DummyDataset(),
+        )
+
+
+@pytest.mark.skipif(reason="Still using DataPipeline Old API")
+@pytest.mark.parametrize("use_preprocess", [False, True])
+@pytest.mark.parametrize("use_postprocess", [False, True])
+def test_data_pipeline_init_and_assignement(use_preprocess, use_postprocess, tmpdir):
+
+    class SubPreprocess(Preprocess):
+        pass
+
+    class SubPostprocess(Postprocess):
+        pass
+
+    data_pipeline = DataPipeline(
+        SubPreprocess() if use_preprocess else None,
+        SubPostprocess() if use_postprocess else None,
+    )
+    assert isinstance(data_pipeline._preprocess_pipeline, SubPreprocess if use_preprocess else Preprocess)
+    assert isinstance(data_pipeline._postprocess_pipeline, SubPostprocess if use_postprocess else Postprocess)
+
+    model = CustomModel(Postprocess())
+    model.data_pipeline = data_pipeline
+    assert isinstance(model._preprocess, Preprocess)
+    assert isinstance(model._postprocess, SubPostprocess if use_postprocess else Postprocess)
+
+
+def test_data_pipeline_is_overriden_and_resolve_function_hierarchy(tmpdir):
+
+    class CustomPreprocess(Preprocess):
+
+        def load_data(self, *_, **__):
+            return 0
+
+        def test_load_data(self, *_, **__):
+            return 1
+
+        def predict_load_data(self, *_, **__):
+            return 2
+
+        def predict_load_sample(self, *_, **__):
+            return 3
+
+        def val_load_sample(self, *_, **__):
+            return 4
+
+        def val_per_sample_pre_tensor_transform(self, *_, **__):
+            return 5
+
+        def predict_per_sample_to_tensor_transform(self, *_, **__):
+            return 7
+
+        def train_per_sample_post_tensor_transform(self, *_, **__):
+            return 8
+
+        def test_collate(self, *_, **__):
+            return 9
+
+        def val_per_sample_transform_on_device(self, *_, **__):
+            return 10
+
+        def train_per_batch_transform_on_device(self, *_, **__):
+            return 11
+
+        def test_per_batch_transform_on_device(self, *_, **__):
+            return 12
```

Comment:
are these numbers checked somehwre?
lest do it rather increment by n in each step and check the final values
```python
def anyme(self, data, ...)
    return data + 3
```

Reply:
removed !